backend: Add Amazon VPC backend

rohansingh opened the pull request:
The Amazon VPC backend adds routes to a VPC route table. Routing is then
handled by Amazon's infrastructure.

rohansingh commented:
Apologies for the out-of-the-blue code drop. It's pretty straightforward though, and the usage is outlined in `README.md`. Let me know if you see any issues, or if this isn't something that you think fits with the direction of flannel.

eyakubovich commented on backend/vpc/vpc.go:
```diff
@@ -0,0 +1,130 @@
+		RouteTableID string
+	}
+	ec2        *ec2.EC2
+	instanceID string
```
Does it need to be a member? I think it can be local var in Init.

eyakubovich commented on backend/vpc/vpc.go:
```diff
@@ -0,0 +1,130 @@
+		Region       string
+		RouteTableID string
+	}
+	ec2        *ec2.EC2
```
Does it need to be a member? I think it can be local var in Init.

rohansingh replied:
Oops. Left over from a draft version.

eyakubovich commented:
This is a great addition. I made some small nits. Also, let's call the backend `aws-vpc` to make it clear that it's for AWS. We've also planned to add the same thing for GCE, so it will help to disambiguate.

rohansingh commented:
@eyakubovich Thanks for the review. Just pushed a version with the following changes:

1. Fixed scoping of the vars you pointed out.
2. Renamed to `aws-vpc`.
3. Auto-detect the region from AWS metadata, so you just specify the route table ID in the config.

eyakubovich commented:
LGTM. Thank you for contributing.

Also, do you know what happens when the traffic is sent in a single VPC subnet? Does it still hop through the router (i.e. extra latency) or is does it stay within the subnet? I understand that this is all virtual so it may all look very different under the hood but it'd be interesting to understand.

benmccann commented:
Nice!  We were just talking about this over at kubernetes where I had some questions about their AWS docs (https://github.com/GoogleCloudPlatform/kubernetes/issues/7036) and Flannel. @brendanburns said "AWS VPC Advanced networking is also capable of subnet per host". I just wanted to clarify that means all AWS VPCs? And not AWS VPCs with "Enhanced Networking"?

This is pretty cool. If the same thing is added for GCE then Flannel could be a pretty generic way of handling this on any cloud and perhaps integrated as a default component of Kubernetes...

rohansingh commented:
@eyakubovich Don't have access to my test rig at the moment, but I'd think traffic within the same subnet wouldn't go through the router. The router would just send packets back to the host on which they originated, after all.

@benmccann The AWS [Enhanced Networking](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/enhanced-networking.html) thing is something else altogether. It's not required for this.

What @bredanburns was referring to is exactly what this backend does, except that the Kube setup scripts do it using the AWS CLI:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/aws/util.sh#L493

By the way, what that setup script is doing with creating a VPC subnet and automatically choosing & associating the correct route table is more sophisticated than what this backend is doing. It might be something to look into emulating.

benmccann commented:
Cool. Thanks for the clarifications!

bakins commented:
Very cool!

Thoughts on using https://github.com/awslabs/aws-sdk-go rather than goamz?  I noticed that Hashicorp has started to use it rather than mitchell's version.  Not 100% sure it supports all the API we need for routes.

benmccann commented:
@justinsb was switching Kubernetes to use awslabs rather than goamz

rohansingh commented:
Looks like it's got support: http://godoc.org/github.com/awslabs/aws-sdk-go/service/ec2#EC2.CreateRoute

justinsb commented:
On switching to the official AWS SDK, I think it is going to happen.  I filed this issue: https://github.com/GoogleCloudPlatform/kubernetes/issues/6162 and @jsdir took up the cause!

Sounds like we're just waiting for a few final things to be supported in the AWS SDK: https://github.com/GoogleCloudPlatform/kubernetes/issues/6162#issuecomment-93910186
